File: src/main/java/spring/webmvc/infrastructure/util/ResponseWriter.java
package spring.webmvc.infrastructure.util;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

import org.springframework.http.MediaType;
import org.springframework.http.ProblemDetail;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.databind.ObjectMapper;

import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class ResponseWriter {

	private final ObjectMapper objectMapper;
	private final HttpServletResponse response;

	public void writeResponse(ProblemDetail problemDetail) throws IOException {
		response.setStatus(problemDetail.getStatus());
		response.setContentType(MediaType.APPLICATION_JSON_VALUE);
		response.setCharacterEncoding(StandardCharsets.UTF_8.toString());
		response.getWriter().write(objectMapper.writeValueAsString(problemDetail));
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/infrastructure/util/ResponseWriter.kt
package spring.web.kotlin.infrastructure.util

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.MediaType
import org.springframework.http.ProblemDetail
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets

@Component
class ResponseWriter(
    private val objectMapper: ObjectMapper,
    private val response: HttpServletResponse
) {
    fun writeResponse(problemDetail: ProblemDetail) {
        response.status = problemDetail.status
        response.contentType = MediaType.APPLICATION_JSON_VALUE
        response.characterEncoding = StandardCharsets.UTF_8.toString()
        response.writer.write(objectMapper.writeValueAsString(problemDetail))
    }
}